076-decouple-ca-management-from-end-entity-certificate-management.md

```diff
@@ -0,0 +1,447 @@
+   This makes it hard to drop in an alternative to the existing Strimzi Cluster CA, since it's behaviour is relied upon throughout the reconcile loop.
+
+This proposal describes changes to make it easier to reason about the current state of CA certificates and track the roll out of trust.
+
+## Proposal
+
+Strimzi should be updated to decouple the management of the cluster CA certificate, and the consumption of the CA certificate by components (such as ZooKeeper, Kafka and cluster operator).
+
+To do this we will introduce a new Kubernetes Secret.
+The new Kubernetes Secret, called <CLUSTER_NAME>-cluster-ca-trusted-certs (referred to from here onwards as trusted-certs Secret) is used to store all Cluster CA certificates that currently need trusting by the cluster and keep track of their current state with respect to issued end entity certificates.
+The trusted-certs Secret is mounted onto Kubernetes Pods and used for their TrustStore.
+This new Kubernetes Secret is separate from the <CLUSTER_NAME>-cluster-ca and <CLUSTER_NAME>-cluster-ca-cert Kubernetes Secrets that are used by Strimzi to store the latest CA certificate and key (or by the user when they are managing their own CA certificates), and by user applications as a Trust Store.
+The cluster operator will update the trusted-certs Kubernetes Secret when it renews the cluster CA certificate (or observes that the user has renewed it).
+
+Introducing this new Kubernetes Secret brings a couple of benefits:
+* The existing Kubernetes Secrets used by the user or Strimzi to store the latest CA certificate (and by user applications as a Trust Store) are no longer directly mounted onto Strimzi's Kubernetes Pods.
+  This allows Strimzi to more carefully manage when the latest CA certificate is trusted by components.
+  If in future components are able to dynamically update their TrustStores, updating the Kubernetes Secret storing the CA certificate doesn't automatically update the CA certificate being trusted.
+  This also means when integrating with a certificate management solution Strimzi doesn't need to reuse these Kubernetes Secrets, it can fetch certificates programmatically for example.
+* It will be easier to reason about which CA certificates are in use/trusted, both as an end user when viewing the cluster, but also as a developer when reviewing the code.
+  This is because currently the state is inferred by observing various Kubernetes annotations and Java class fields.
+
+These benefits will also make it easier to evolve the management of CA certificates in future (see [Future changes](#future-changes)).
+
+Non-goals of the proposal:
+* This proposal does not change the management of Clients CA.
+  Users can provide custom listener certificates, so allowing integration with a certificate management solution for the Clients CA doesn't bring as many benefits as it does for the Cluster CA.
```

Comment:
I don't understand this:
* The custom listener certificates have nothing to do with Cluster CA
* There is a demand (I would say equally big as for Cluster CA if not higher, although the challenges are possibly a bit different) for integrating the Clients CA with external certificate management

Reply:
> The custom listener certificates have nothing to do with Cluster CA

In a previous version of the proposal I was also proposing changes to the Clients CA (which is related to custom listener certificates), however I removed that to focus on the Cluster CA and reduce the scope of the proposal. The hope is that if the proposal is narrowly scoped it's easier to get agreement and make progress.

> There is a demand (I would say equally big as for Cluster CA if not higher, although the challenges are possibly a bit different) for integrating the Clients CA with external certificate management

I'm not saying there isn't demand. However for the Cluster CA today regardless of how you configure Strimzi you have to have access to the CA private key so you can give that to Strimzi. Whereas that isn't the case for the Clients CA if you provide custom listener certificates. So that is the key point I was thinking of here and why I started with the Cluster CA not the Clients CA. I can add that clarification explicitly in the proposal.